ykrt/src/location.rs

```diff
@@ -62,39 +69,22 @@ pub struct Location {
-    //
-    // The state machine is encoded in a `usize` in a not-entirely-simple way, as we don't want to
-    // allocate any memory for Locations that do not become hot. The layout is as follows (on a 64
-    // bit machine):
+    // there. However, many Locations will not be used frequently enough to reach such a state, so
+    // we don't want to waste resources on them.
     //
-    //   bit(s) | 63..3   | 2           | 1         | 0
-    //          | payload | IS_COUNTING | IS_PARKED | IS_LOCKED
+    // We therefore encode a Location as a tagged integer: in the initial (Counting) state, no
+    // memory is allocated; if the location is used frequently enough it becomes hot, memory
+    // is allocated for it, and a pointer stored instead of an integer. Note that once memory for a
+    // hot location is allocated, it can only be (scheduled for) deallocation when a Location
+    // is dropped, as the Location may have handed out `&` references to that allocated memory.
     //
-    // In the `Counting` state, `IS_COUNTING` is set to 1, and `IS_PARKED` and `IS_LOCKED` are
-    // unused and must remain set at 0. The payload representing the count is incremented
-    // locklessly. All other states have `IS_COUNTING` set to 0 and the payload is the address of a
-    // `Box<HotLocation>`, access to which is controlled by `IS_LOCKED`.
+    // The layout of a Location is as follows (on a 64 bit machine):
     //
-    // The possible combinations of the counting and mutex bits are thus as follows:
+    //   bit(s) | 63..1   | 0
+    //          | payload | <STATE_NOT_HOT|STATE_HOT>
     //
-    //   payload          | IS_COUNTING | IS_PARKED | IS_LOCKED | Notes
-    //   -----------------+-----------+---------+---------+-----------------------------------
-    //   count            | 1           | 0         | 0         | Start state
-    //                    | 1           | 0         | 1         | Illegal state
-    //                    | 1           | 1         | 0         | Illegal state
-    //                    | 1           | 1         | 1         | Illegal state
-    //   Box<HotLocation> | 0           | 0         | 0         | Not locked, no-one waiting
-    //   Box<HotLocation> | 0           | 0         | 1         | Locked, no thread(s) waiting
-    //   Box<HotLocation> | 0           | 1         | 0         | Not locked, thread(s) waiting
-    //   Box<HotLocation> | 0           | 1         | 1         | Locked, thread(s) waiting
-    //
-    // where `count` is an integer and `Box<HotLocation>` is a pointer to a `malloc`ed block of
-    // memory in the heap containing a `HotLocation` enum.
-    //
-    // The precise semantics of locking and, in particular, parking are subtle: interested readers
-    // are directed to https://github.com/Amanieu/parking_lot/blob/master/src/raw_mutex.rs#L33 for
-    // a more precise definition.
+    // In the `STATE_NOT_HOT` state, the payload is an integer; in a `STATE_HOT` state, the payload
```

Comment:
What's the content of the payload in `STATE_NOT_HOT`? I assumed from the description above, that a `Location` has two states: a counter or a pointer. But it seems there's a third that I don't see.

Reply:
Looking at this more I might just be misinterpreting this table:
```
    //   bit(s) | 63..1   | 0
    //          | payload | <STATE_NOT_HOT|STATE_HOT>
```

What does the first row tell me?

Reply:
Another way of reading this is "bit 0 == <STATE_NOT_HOT|STATE_HOT>, bits 1..63 = <payload>". Perhaps the table is not as useful as when there were multiple possibilities here. I can certainly get rid of it.

Reply:
Oh! I see now.  I thought the two rows were different options. Maybe we can change this to
```
    //   bit(s) | 63..1   | 0
    // -------- | ------- | ---
    //          | payload | <STATE_NOT_HOT|STATE_HOT>
```
to make clearer the first column is the title header?

Reply:
I just inlined it into prose in https://github.com/ykjit/yk/commit/7c485783850902032c0532e5d30e11fd76b162e0 if that works for you?

Reply:
Yeah, that works!